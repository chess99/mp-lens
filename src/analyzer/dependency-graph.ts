/**
 * 依赖图数据结构
 * 用于存储文件之间的依赖关系，支持节点和边的操作
 */
export class DependencyGraph {
  private _nodes: Set<string> = new Set();
  private _outEdges: Map<string, Set<string>> = new Map();
  private _inEdges: Map<string, Set<string>> = new Map();

  /**
   * 向图中添加节点
   * @param node 节点名称（文件路径）
   */
  addNode(node: string): void {
    this._nodes.add(node);
    
    // 确保每个节点都有对应的边集合
    if (!this._outEdges.has(node)) {
      this._outEdges.set(node, new Set());
    }
    
    if (!this._inEdges.has(node)) {
      this._inEdges.set(node, new Set());
    }
  }

  /**
   * 向图中添加边（依赖关系）
   * @param from 源文件
   * @param to 目标文件（被依赖）
   */
  addEdge(from: string, to: string): void {
    // 确保两个节点都存在
    this.addNode(from);
    this.addNode(to);
    
    // 添加边
    this._outEdges.get(from)!.add(to);
    this._inEdges.get(to)!.add(from);
  }

  /**
   * 获取所有节点
   */
  nodes(): string[] {
    return [...this._nodes];
  }

  /**
   * 获取节点的所有出边（依赖）
   * @param node 节点名称
   */
  outEdges(node: string): string[] {
    if (!this._outEdges.has(node)) {
      return [];
    }
    return [...this._outEdges.get(node)!];
  }

  /**
   * 获取节点的所有入边（被谁依赖）
   * @param node 节点名称
   */
  inEdges(node: string): string[] {
    if (!this._inEdges.has(node)) {
      return [];
    }
    return [...this._inEdges.get(node)!];
  }

  /**
   * 获取节点的出度（依赖了多少文件）
   * @param node 节点名称
   */
  outDegree(node: string): number {
    if (!this._outEdges.has(node)) {
      return 0;
    }
    return this._outEdges.get(node)!.size;
  }

  /**
   * 获取节点的入度（被多少文件依赖）
   * @param node 节点名称
   */
  inDegree(node: string): number {
    if (!this._inEdges.has(node)) {
      return 0;
    }
    return this._inEdges.get(node)!.size;
  }

  /**
   * 检查图中是否存在节点
   * @param node 节点名称
   */
  hasNode(node: string): boolean {
    return this._nodes.has(node);
  }

  /**
   * 检查图中是否存在从from到to的边
   * @param from 源节点
   * @param to 目标节点
   */
  hasEdge(from: string, to: string): boolean {
    if (!this._outEdges.has(from)) {
      return false;
    }
    return this._outEdges.get(from)!.has(to);
  }

  /**
   * 获取图中的节点数量
   */
  get nodeCount(): number {
    return this._nodes.size;
  }

  /**
   * 获取图中的边数量
   */
  get edgeCount(): number {
    let count = 0;
    for (const edges of this._outEdges.values()) {
      count += edges.size;
    }
    return count;
  }

  /**
   * 将图转换为JSON格式
   */
  toJSON() {
    const nodes = this.nodes();
    const links = nodes.flatMap(from => 
      this.outEdges(from).map(to => ({ source: from, target: to }))
    );
    
    return {
      nodes: nodes.map(id => ({ id })),
      links
    };
  }
} 