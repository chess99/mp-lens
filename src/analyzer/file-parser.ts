import * as fs from 'fs';
import * as path from 'path';
import { AnalyzerOptions } from '../types/command-options';
import { AliasResolver } from '../utils/alias-resolver';
import { logger } from '../utils/debug-logger';
import { PathResolver } from './utils/path-resolver';

// Import specialized parsers with corrected paths relative to src/analyzer/
import { JavaScriptParser } from './parsers/javascript-parser';
import { JSONParser } from './parsers/json-parser';
import { WXMLParser } from './parsers/wxml-parser';
import { WXSSParser } from './parsers/wxss-parser';

/**
 * FileParser: Orchestrates parsing of different file types in a WeChat Mini Program.
 * Handles file reading and path resolution, then delegates text analysis to specialized parsers.
 */
export class FileParser {
  private pathResolver: PathResolver;

  // Store instances of specialized parsers
  private javaScriptParser: JavaScriptParser;
  private wxmlParser: WXMLParser;
  private wxssParser: WXSSParser;
  private jsonParser: JSONParser;

  constructor(projectRoot: string, options: AnalyzerOptions) {
    const actualRoot = options.miniappRoot || projectRoot;

    if (options.miniappRoot) {
      logger.debug(`FileParser using custom miniapp root: ${options.miniappRoot}`);
    }

    const aliasResolver = new AliasResolver(actualRoot);
    const hasAliasConfig = aliasResolver.initialize();

    if (hasAliasConfig) {
      logger.debug('Alias configuration detected, automatically enabling alias resolution');
      logger.debug('Alias configuration:', aliasResolver.getAliases());
    }

    // Instantiate PathResolver
    this.pathResolver = new PathResolver(projectRoot, options, aliasResolver, hasAliasConfig);

    // Instantiate specialized parsers - they now only handle text analysis
    this.javaScriptParser = new JavaScriptParser();
    this.wxmlParser = new WXMLParser();
    this.wxssParser = new WXSSParser();
    this.jsonParser = new JSONParser();
  }

  /**
   * Parses a single file by reading its content and delegating text analysis to the appropriate parser.
   * Handles path resolution centrally.
   * Returns a list of absolute paths to the file's dependencies.
   */
  async parseFile(filePath: string): Promise<string[]> {
    const ext = path.extname(filePath).toLowerCase();

    try {
      // Read file content once at the top level
      const content = fs.readFileSync(filePath, 'utf-8');
      let rawDependencies: string[] = [];

      // Delegate text analysis to specialized parsers
      switch (ext) {
        case '.js':
        case '.ts':
        case '.wxs': // WXS files are JavaScript, use the same parser
          rawDependencies = await this.javaScriptParser.parse(content, filePath);
          break;
        case '.wxml':
          rawDependencies = await this.wxmlParser.parse(content, filePath);
          break;
        case '.wxss':
          rawDependencies = await this.wxssParser.parse(content, filePath);
          break;
        case '.json':
          rawDependencies = await this.jsonParser.parse(content, filePath);
          break;
        case '.png':
        case '.jpg':
        case '.jpeg':
        case '.gif':
        case '.svg':
          return []; // Image files have no dependencies
        default:
          logger.trace(`Unsupported file type for parsing: ${filePath}`);
          return [];
      }

      // Resolve all raw dependency paths to absolute paths
      const resolvedDependencies: string[] = [];
      for (const rawPath of rawDependencies) {
        const resolvedPath = this.resolveDependencyPath(rawPath, filePath, ext);
        if (resolvedPath) {
          resolvedDependencies.push(resolvedPath);
        }
      }

      return resolvedDependencies;
    } catch (e: any) {
      // Centralized error handling for file reading or parsing issues
      logger.warn(`Error parsing file ${filePath}: ${e.message}`);
      return []; // Return empty array on error
    }
  }

  /**
   * Resolves a raw dependency path to an absolute path based on file type context
   */
  private resolveDependencyPath(
    rawPath: string,
    sourcePath: string,
    sourceExt: string,
  ): string | null {
    // Determine allowed extensions based on source file type and dependency context
    let allowedExtensions: string[];

    switch (sourceExt) {
      case '.js':
      case '.ts':
        allowedExtensions = ['.js', '.ts', '.json'];
        break;
      case '.wxs':
        allowedExtensions = ['.wxs']; // WXS files can only import other WXS files
        break;
      case '.wxml':
        // WXML can reference .wxml (import/include), .wxs, and image files
        // Determine type based on path characteristics
        if (this.isImagePath(rawPath)) {
          allowedExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp'];
        } else if (rawPath.includes('.wxs') || rawPath.endsWith('.wxs')) {
          allowedExtensions = ['.wxs'];
        } else {
          // Default to WXML for import/include
          allowedExtensions = ['.wxml'];
        }
        break;
      case '.wxss':
        // WXSS can import other WXSS files or reference image files
        if (this.isImagePath(rawPath)) {
          allowedExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp'];
        } else {
          allowedExtensions = ['.wxss'];
        }
        break;
      case '.json':
        // JSON files can reference various file types depending on context
        // For pages and components, we need to find all related files
        if (this.isImagePath(rawPath)) {
          allowedExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp'];
        } else {
          // For pages/components, try to find the main file first
          allowedExtensions = ['.js', '.ts', '.wxml', '.wxss', '.json'];
        }
        break;
      default:
        allowedExtensions = [];
    }

    return this.pathResolver.resolveAnyPath(rawPath, sourcePath, allowedExtensions);
  }

  /**
   * Determines if a path refers to an image file based on its extension
   */
  private isImagePath(filePath: string): boolean {
    const imageExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp'];
    const ext = path.extname(filePath).toLowerCase();

    // Strictly check if the file extension is an image extension
    return imageExtensions.includes(ext);
  }
}
