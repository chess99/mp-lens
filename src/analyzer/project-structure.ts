export type NodeType = 'App' | 'Package' | 'Page' | 'Component' | 'Module';
export type LinkType =
  | 'Structure' // Hierarchical relationship between components (App->Pages->Components) defined in configuration
  | 'Import'
  | 'Style'
  | 'Template'
  | 'Config' // Config for page/component -> json, App -> app.json
  | 'Resource' // Link to assets like images (e.g., from tabBar)
  | 'WorkerEntry'; // Link from App to worker entry point

export interface GraphNode {
  id: string; // File path or logical identifier (e.g., 'app', 'pkg:subPackageRoot')
  type: NodeType;
  label: string; // User-friendly name or path relative to miniapp root
  // Optional metadata
  properties?: {
    // Common properties
    absolutePath?: string; // For Module nodes
    basePath?: string; // For Page/Component nodes
    root?: string; // For Package nodes
    path?: string; // For App node (path to app.json)
    // File properties (for Module)
    fileSize?: number;
    fileExt?: string;
    // Statistics properties (calculated in UI, not populated by backend)
    fileCount?: number;
    totalSize?: number;
    fileTypes?: Record<string, number>; // { ext: count }
    sizeByType?: Record<string, number>; // { ext: size }
    // --- NEW PROPERTIES ---
    structuralParentId?: string; // ID of the Page/Component/Package this Module primarily belongs to
    referredBy?: string[]; // List of node IDs that import/require this node (populated during parsing)
    // --- END NEW PROPERTIES ---
    // Add other relevant properties as needed
    [key: string]: any; // Allow other properties
  };
}

export interface GraphLink {
  source: string; // ID of the source node
  target: string; // ID of the target node
  type: LinkType; // Type of the relationship
  dependencyType?: string; // Specific type of dependency (e.g., 'static', 'dynamic')
  // Optional metadata
  properties?: Record<string, any>;
}

export interface ProjectStructure {
  nodes: GraphNode[];
  links: GraphLink[];
  rootNodeId: string | null; // ID of the 'App' node
  miniappRoot: string; // Absolute path to miniapp root
}

// Maybe add helper functions here later, e.g., findNodeById, findLinksFrom, etc.
