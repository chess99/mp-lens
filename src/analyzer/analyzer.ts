import * as fs from 'fs';
import * as glob from 'glob';
import * as path from 'path';
import { AnalyzerOptions } from '../types/command-options';
import { logger } from '../utils/debug-logger';
import { HandledError } from '../utils/errors';
import { findPureAmbientDeclarationFiles } from '../utils/typescript-helper';
import { GraphLink, ProjectStructure } from './project-structure';
import { ProjectStructureBuilder } from './project-structure-builder';

interface AnalysisResult {
  projectStructure: ProjectStructure;
  unusedFiles: string[];
  reachableNodeIds: Set<string>;
}

// --- Start: Helper Function Definitions --- //

/**
 * Finds all .d.ts files that are pure ambient declarations.
 */
// function findPureAmbientDeclarationFiles(projectRoot: string, allFiles: string[]): string[] {
// Implementation might be complex, assume it exists and works
// return [];
// }

/**
 * Resolves user-provided essential file paths relative to project/miniapp roots.
 */
// function resolveEssentialFiles(
//   projectRoot: string,
//   miniappRoot: string,
//   essentialFilesUser: string[],
// ): Set<string> {
// Implementation needs file system access
// return new Set<string>();
// }

/**
 * Performs reachability analysis (BFS) on the ProjectStructure graph.
 */
function findReachableNodes(structure: ProjectStructure, entryNodeIds: string[]): Set<string> {
  const reachable = new Set<string>();
  const queue: string[] = [];
  const linksFrom = new Map<string, GraphLink[]>();
  const nodeMap = new Map(structure.nodes.map((n) => [n.id, n]));

  // Precompute outgoing links for faster lookup
  structure.links.forEach((link) => {
    if (!linksFrom.has(link.source)) {
      linksFrom.set(link.source, []);
    }
    linksFrom.get(link.source)!.push(link);
  });

  // Initialize queue with entry points that exist in the graph
  entryNodeIds.forEach((id) => {
    if (nodeMap.has(id)) {
      queue.push(id);
      reachable.add(id);
    } else {
      logger.warn(`Entry node ID specified but not found in graph: ${id}`);
    }
  });

  logger.trace(`findReachableNodes: Starting BFS with initial queue size ${queue.length}`);
  while (queue.length > 0) {
    const currentNodeId = queue.shift()!;
    logger.trace(`BFS: Processing node ${nodeMap.get(currentNodeId)?.label} [${currentNodeId}]`);

    // Find nodes reachable from the current node
    const outgoingLinks = linksFrom.get(currentNodeId) || [];
    logger.trace(`BFS: Found ${outgoingLinks.length} outgoing links from ${currentNodeId}`);

    for (const link of outgoingLinks) {
      const targetNodeId = link.target;
      const targetNodeLabel = nodeMap.get(targetNodeId)?.label;
      logger.trace(
        `BFS: Checking neighbor ${targetNodeLabel} [${targetNodeId}] via link type ${link.type}`,
      );
      if (nodeMap.has(targetNodeId) && !reachable.has(targetNodeId)) {
        reachable.add(targetNodeId);
        queue.push(targetNodeId);
        logger.trace(`BFS: Added ${targetNodeLabel} [${targetNodeId}] to reachable set and queue.`);
      } else if (!nodeMap.has(targetNodeId)) {
        logger.trace(`BFS: Neighbor node not found in map: ${targetNodeId}`);
      } else if (reachable.has(targetNodeId)) {
        logger.trace(`BFS: Neighbor already reachable: ${targetNodeLabel} [${targetNodeId}]`);
      }
    }
  }

  logger.debug(`Reachability analysis complete. Found ${reachable.size} reachable nodes.`);
  return reachable;
}

// --- End: New Helper Function --- //

// --- Start: Re-added findAllFiles --- //
/**
 * 在指定目录中查找所有符合条件的文件
 */
function findAllFiles(rootDir: string, fileTypes: string[], excludePatterns: string[]): string[] {
  // Ensure fileTypes are valid for glob pattern
  const safeFileTypes = fileTypes.filter((t) => t && /^[a-zA-Z0-9]+$/.test(t));
  if (safeFileTypes.length === 0) {
    logger.warn('No valid file types specified for glob search.');
    return [];
  }
  const globPattern = `**/*.{${safeFileTypes.join(',')}}`;

  // Default ignore patterns (consider refining these)
  const defaultIgnorePatterns = [
    '**/node_modules/**',
    '**/miniprogram_npm/**',
    // Add other common build/output dirs if necessary, e.g., '**/dist/**'?
    // Keep output patterns from original
    '**/output/dependency-graph.*',
    '**/output/unused-files.*',
    'dependency-graph.*',
    'unused-files.*',
  ];

  const globOptions = {
    cwd: rootDir,
    absolute: true,
    ignore: [...defaultIgnorePatterns, ...excludePatterns],
    nodir: true,
    dot: true, // Include hidden files/folders if not excluded
  };

  logger.debug(`Glob pattern: ${globPattern}`);
  logger.debug(`Glob options:`, globOptions);

  try {
    const files = glob.sync(globPattern, globOptions);
    logger.debug(`Found ${files.length} files via initial scan in ${rootDir}`);
    return files;
  } catch (error) {
    logger.error(`Error during initial file scan:`, error);
    return [];
  }
}
// --- End: Re-added findAllFiles --- //

// --- Start: Exported analyzeProject function --- //

export async function analyzeProject(
  projectRoot: string,
  options: AnalyzerOptions,
): Promise<AnalysisResult> {
  const {
    fileTypes = ['js', 'ts', 'wxml', 'wxss', 'json'],
    excludePatterns = [],
    miniappRoot,
    appJsonPath,
    appJsonContent,
    essentialFiles = [],
    includeAssets = false,
  } = options;

  logger.debug('Project Root:', projectRoot);
  logger.debug('MiniApp Root:', miniappRoot);
  logger.debug('Exclude patterns:', excludePatterns);
  logger.debug('Include assets in cleanup:', includeAssets);
  logger.debug('Essential files:', essentialFiles);

  // 验证 appJsonContent 存在
  if (
    !appJsonContent ||
    typeof appJsonContent !== 'object' ||
    Object.keys(appJsonContent).length === 0
  ) {
    const errorMsg =
      '分析失败: 没有找到有效的 app.json 内容。请确保小程序项目根目录中存在 app.json 文件，或通过配置提供 appJsonContent。';
    logger.error(errorMsg);
    throw new HandledError(errorMsg);
  }

  // --- Initial File Scan --- //
  // Scan should happen within the miniapp root if specified
  const allFoundFiles = findAllFiles(miniappRoot, fileTypes, excludePatterns);
  if (allFoundFiles.length === 0) {
    // If no files found, analysis will be based solely on appJsonContent
    logger.warn('在指定的 miniapp 根目录中未找到匹配的文件。');
  }

  // Add app.json path if it exists and wasn't found by glob (e.g., different extension)
  if (appJsonPath && !allFoundFiles.includes(appJsonPath)) {
    logger.debug(`Manually adding app.json path to found files: ${appJsonPath}`);
    allFoundFiles.push(appJsonPath);
  }

  // --- Build Project Structure ---
  const builder = new ProjectStructureBuilder(
    projectRoot,
    miniappRoot,
    appJsonPath,
    appJsonContent,
    allFoundFiles, // Pass the list of all files
    options, // Pass remaining options
  );

  const projectStructure = await builder.build();

  // --- Calculate Reachable Nodes FIRST --- //
  const nodeMap = new Map(projectStructure.nodes.map((n) => [n.id, n]));
  const entryNodeIdsSet: Set<string> = new Set();
  if (projectStructure.rootNodeId) {
    entryNodeIdsSet.add(projectStructure.rootNodeId);
  } else {
    logger.warn('Project structure has no root node ID defined.');
  }
  // Add essential files as entry points
  const essentialFilePaths = resolveEssentialFiles(projectRoot, miniappRoot, essentialFiles);
  essentialFilePaths.forEach((filePath) => {
    if (nodeMap.has(filePath)) {
      entryNodeIdsSet.add(filePath);
      logger.trace(`Added essential file path as entry point: ${filePath}`);
    } else {
      logger.trace(`Essential file path specified but not found as node: ${filePath}`);
    }
  });
  // Add pure ambient declaration files as entry points
  const allFilePathsInStructure = projectStructure.nodes
    .filter((n) => n.type === 'Module' && n.properties?.absolutePath)
    .map((n) => n.properties!.absolutePath as string);
  const pureAmbientDtsFiles = findPureAmbientDeclarationFiles(projectRoot, allFilePathsInStructure);
  pureAmbientDtsFiles.forEach((filePath) => {
    if (nodeMap.has(filePath)) {
      entryNodeIdsSet.add(filePath);
      logger.trace(`Added pure ambient d.ts as entry point: ${filePath}`);
    } else {
      logger.warn(`Ambient d.ts file found but not present as node in structure: ${filePath}`);
    }
  });
  logger.debug(`Starting reachability analysis from ${entryNodeIdsSet.size} entry nodes.`);
  const reachableNodeIds = findReachableNodes(projectStructure, Array.from(entryNodeIdsSet));
  logger.debug(`Found ${reachableNodeIds.size} total reachable nodes.`);

  // --- Find Unused Files using the Calculated Reachable Nodes --- //
  const unusedFiles = findUnusedFiles(
    projectStructure,
    projectRoot,
    reachableNodeIds, // <-- Pass calculated reachable nodes
    includeAssets,
  );

  return {
    projectStructure,
    unusedFiles,
    reachableNodeIds, // <-- Return calculated reachable nodes
  };
}

/**
 * Finds unused files using the ProjectStructure.
 * Relies on pre-calculated reachableNodeIds.
 */
function findUnusedFiles(
  structure: ProjectStructure,
  projectRoot: string,
  reachableNodeIds: Set<string>, // <-- Receive reachable nodes
  includeAssets: boolean = false,
): string[] {
  // --- Step 1: Determine Unused FILES --- //
  const allModuleNodes = structure.nodes.filter((n) => n.type === 'Module');
  const unusedModuleNodes = allModuleNodes.filter((node) => !reachableNodeIds.has(node.id));
  let unusedFiles = unusedModuleNodes
    .map((node) => node.properties?.absolutePath as string | undefined)
    .filter((filePath): filePath is string => !!filePath);

  // --- Step 2: Filter out asset files unless includeAssets is true --- //
  if (!includeAssets) {
    const assetExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg'];
    logger.debug(`Filtering out asset files with extensions: ${assetExtensions.join(', ')}`);

    unusedFiles = unusedFiles.filter((absolutePath) => {
      const fileExt = path.extname(absolutePath).toLowerCase();
      const isAsset = assetExtensions.includes(fileExt);

      if (isAsset) {
        const relativePath = path.relative(projectRoot, absolutePath);
        logger.verbose(`Keeping asset file (assets are excluded by default): ${relativePath}`);
      }

      // Keep the file in the unused list only if it's NOT an asset or if we explicitly want to include assets
      return !isAsset;
    });
  }

  // Debugging output (adjusted)
  logger.verbose('Total Module Nodes:', allModuleNodes.length);
  // logger.verbose('Entry Node IDs for BFS:', entryNodeIds.size); // Removed, happens before
  logger.verbose('Reachable Node IDs:', reachableNodeIds.size);
  logger.info(`发现 ${unusedFiles.length} 个未使用的文件。`);

  return unusedFiles;
}

/**
 * Resolves essential files provided by the user to absolute paths.
 * (Implementation likely remains the same - it deals with file paths)
 */
function resolveEssentialFiles(
  projectRoot: string,
  miniappRoot: string,
  essentialFilesUser: string[],
): Set<string> {
  const essentialFiles = new Set<string>();

  // --- Start: Added Default Essential Files ---
  const projectLevelFiles = [
    'tsconfig.json',
    'jsconfig.json',
    'mp-lens.config.json',
    'mp-lens.config.js',
    'package.json',
    '.eslintrc.js',
    '.eslintrc.json',
    '.prettierrc',
    '.prettierrc.js',
    '.babelrc',
    'babel.config.js',
  ];

  const miniappLevelFiles = [
    'app.json', // Often the source of truth for entries, essential itself
    'project.config.json',
    'project.private.config.json',
    'sitemap.json',
    'theme.json',
    'ext.json',
  ];

  // Add default project-level files
  projectLevelFiles.forEach((file) => {
    const absPath = path.resolve(projectRoot, file);
    // Check existence? Optional, depends if we want to add only existing ones.
    // Adding regardless simplifies logic, reachability check will handle non-existent nodes later.
    // if (fs.existsSync(absPath)) {
    essentialFiles.add(absPath);
    // }
  });

  // Add default miniapp-level files
  miniappLevelFiles.forEach((file) => {
    const absPath = path.resolve(miniappRoot, file);
    // if (fs.existsSync(absPath)) {
    essentialFiles.add(absPath);
    // }
  });
  logger.verbose(
    `Added ${projectLevelFiles.length + miniappLevelFiles.length} default essential file paths.`,
  );
  // --- End: Added Default Essential Files ---

  // Add user-specified essential files
  essentialFilesUser.forEach((file) => {
    // Try resolving relative to miniapp root first, then project root
    const pathFromMiniApp = path.resolve(miniappRoot, file);
    const pathFromProject = path.resolve(projectRoot, file);

    if (fs.existsSync(pathFromMiniApp)) {
      essentialFiles.add(pathFromMiniApp);
      logger.verbose(`Resolved essential file (from miniapp root): ${file} -> ${pathFromMiniApp}`);
    } else if (fs.existsSync(pathFromProject)) {
      essentialFiles.add(pathFromProject);
      logger.verbose(`Resolved essential file (from project root): ${file} -> ${pathFromProject}`);
    } else {
      logger.warn(`Specified essential file not found: ${file}`);
    }
  });
  return essentialFiles;
}
