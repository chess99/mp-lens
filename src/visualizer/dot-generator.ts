import * as path from 'path';
import { DependencyGraph } from '../analyzer/dependency-graph';

interface DotGeneratorOptions {
  title: string;
  projectRoot: string;
  maxDepth?: number;
  focusNode?: string;
}

/**
 * DOT 依赖图生成器
 * 生成 Graphviz 兼容的 DOT 格式
 */
export class DotGenerator {
  private graph: DependencyGraph;

  constructor(graph: DependencyGraph) {
    this.graph = graph;
  }

  /**
   * 生成 DOT 格式的依赖图
   */
  generate(options: DotGeneratorOptions): string {
    const { title, projectRoot, maxDepth, focusNode } = options;

    // 准备节点和边的数据
    const { nodes, edges } = this.prepareGraphData(projectRoot, maxDepth, focusNode);

    // 构建 DOT 字符串
    let dot = `digraph "${title}" {\n`;

    // 图的全局设置
    dot += '  graph [rankdir=LR, fontname="Arial", fontsize=12, overlap=false, splines=true];\n';
    dot += '  node [shape=box, style="rounded,filled", fontname="Arial", fontsize=10];\n';
    dot += '  edge [color="#999999", fontname="Arial", fontsize=8];\n\n';

    // 添加节点
    for (const node of nodes) {
      const attrs = this.getNodeAttributes(node);

      // 格式化属性
      const attrStr = Object.entries(attrs)
        .map(([key, value]) => `${key}="${value}"`)
        .join(', ');

      dot += `  "${node.id}" [${attrStr}];\n`;
    }

    dot += '\n';

    // 添加边
    for (const edge of edges) {
      const attrs = this.getEdgeAttributes(edge);

      // 格式化属性
      const attrStr = Object.entries(attrs)
        .map(([key, value]) => `${key}="${value}"`)
        .join(', ');

      dot += `  "${edge.source}" -> "${edge.target}" [${attrStr}];\n`;
    }

    dot += '}\n';

    return dot;
  }

  /**
   * 准备用于生成 DOT 的图形数据
   */
  private prepareGraphData(projectRoot: string, maxDepth?: number, focusNode?: string) {
    // 获取所有节点
    const allNodes = this.graph.nodes();

    // 结果容器
    const result = {
      nodes: [] as any[],
      edges: [] as any[],
    };

    // 如果指定了焦点节点和最大深度
    if (focusNode && maxDepth !== undefined) {
      // 使用 BFS 遍历图进行筛选
      const includedNodes = new Set<string>();
      const queue: Array<{ node: string; depth: number }> = [];

      // 添加焦点节点
      includedNodes.add(focusNode);
      queue.push({ node: focusNode, depth: 0 });

      // BFS 遍历
      while (queue.length > 0) {
        const { node, depth } = queue.shift()!;

        // 如果达到最大深度，则不继续遍历
        if (depth >= maxDepth) continue;

        // 添加所有出边相连的节点
        for (const target of this.graph.outEdges(node)) {
          if (!includedNodes.has(target)) {
            includedNodes.add(target);
            queue.push({ node: target, depth: depth + 1 });
          }
        }

        // 添加所有入边相连的节点
        for (const source of this.graph.inEdges(node)) {
          if (!includedNodes.has(source)) {
            includedNodes.add(source);
            queue.push({ node: source, depth: depth + 1 });
          }
        }
      }

      // 根据筛选结果构建图形数据
      for (const node of includedNodes) {
        result.nodes.push(this.createNodeObject(node, projectRoot, node === focusNode));
      }

      // 添加边
      for (const source of includedNodes) {
        for (const target of this.graph.outEdges(source)) {
          if (includedNodes.has(target)) {
            result.edges.push({
              source,
              target,
              highlighted: source === focusNode || target === focusNode,
            });
          }
        }
      }
    } else {
      // 没有焦点节点或深度限制，使用全部图
      result.nodes = allNodes.map((node) =>
        this.createNodeObject(node, projectRoot, focusNode === node),
      );

      // 添加边
      for (const source of allNodes) {
        for (const target of this.graph.outEdges(source)) {
          result.edges.push({
            source,
            target,
            highlighted: focusNode && (source === focusNode || target === focusNode),
          });
        }
      }
    }

    return result;
  }

  /**
   * 创建节点对象
   */
  private createNodeObject(nodePath: string, projectRoot: string, highlighted: boolean) {
    const relativePath = path.relative(projectRoot, nodePath);
    const ext = path.extname(nodePath);

    // 确定节点类型
    let type = '';
    if (nodePath.includes('/components/') || nodePath.includes('\\components\\')) {
      type = 'component';
    } else if (nodePath.includes('/pages/') || nodePath.includes('\\pages\\')) {
      type = 'page';
    } else if (ext === '.wxs') {
      type = 'wxs';
    }

    return {
      id: nodePath,
      label: relativePath,
      type,
      highlighted,
    };
  }

  /**
   * 获取节点的 DOT 属性
   */
  private getNodeAttributes(node: any) {
    const attrs: Record<string, string> = {
      label: node.label,
    };

    // 根据类型设置不同的样式
    switch (node.type) {
      case 'component':
        attrs.fillcolor = '#c2f5e2';
        attrs.tooltip = `组件: ${node.label}`;
        break;
      case 'page':
        attrs.fillcolor = '#f5c2dd';
        attrs.tooltip = `页面: ${node.label}`;
        break;
      case 'wxs':
        attrs.fillcolor = '#f5ddc2';
        attrs.tooltip = `WXS模块: ${node.label}`;
        break;
      default:
        attrs.fillcolor = '#c2d1f5';
        attrs.tooltip = `文件: ${node.label}`;
    }

    // 如果是高亮节点，使用不同的颜色
    if (node.highlighted) {
      attrs.fillcolor = '#ff9999';
      attrs.penwidth = '2';
    }

    return attrs;
  }

  /**
   * 获取边的 DOT 属性
   */
  private getEdgeAttributes(edge: any) {
    const attrs: Record<string, string> = {};

    if (edge.highlighted) {
      attrs.color = '#ff0000';
      attrs.penwidth = '1.5';
    }

    return attrs;
  }
}
