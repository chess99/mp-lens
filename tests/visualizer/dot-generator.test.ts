import { GraphLink, GraphNode, ProjectStructure } from '../../src/analyzer/project-structure';
import { DotGenerator, DotGeneratorOptions } from '../../src/visualizer/dot-generator';

describe('DotGenerator', () => {
  it('should generate basic DOT output', () => {
    // Simple mock structure
    const nodes: GraphNode[] = [
      { id: 'app', type: 'App', label: 'App' },
      { id: 'pageA', type: 'Page', label: 'PageA' },
      { id: 'compB', type: 'Component', label: 'CompB' },
      { id: 'utilC', type: 'Module', label: 'UtilC' },
    ];
    const links: GraphLink[] = [
      { source: 'app', target: 'pageA', type: 'Structure' },
      { source: 'pageA', target: 'compB', type: 'Structure' },
      { source: 'pageA', target: 'utilC', type: 'Import' },
      { source: 'compB', target: 'utilC', type: 'Import' },
    ];
    const structure: ProjectStructure = {
      rootNodeId: 'app',
      nodes,
      links,
      miniappRoot: '/test',
    };

    const generator = new DotGenerator(structure);
    const options: DotGeneratorOptions = { title: 'Test Graph' };
    const dotOutput = generator.generate(options);

    // Basic checks
    expect(dotOutput).toContain('digraph "Test Graph"');
    // Remove basic contains checks for nodes/edges as attributes make them brittle
    // expect(dotOutput).toContain('"app" [label="App"];');
    // expect(dotOutput).toContain('"pageA" [label="PageA"];');
    // expect(dotOutput).toContain('"compB" [label="CompB"];');
    // expect(dotOutput).toContain('"utilC" [label="UtilC"];');
    // expect(dotOutput).toContain('"app" -> "pageA"');
    // expect(dotOutput).toContain('"pageA" -> "compB"');
    // expect(dotOutput).toContain('"pageA" -> "utilC"');
    // expect(dotOutput).toContain('"compB" -> "utilC"');

    // Snapshot test for detailed structure/attributes
    expect(dotOutput).toMatchSnapshot();
  });

  // Add more tests for options like depth, focus, etc. if needed
});
