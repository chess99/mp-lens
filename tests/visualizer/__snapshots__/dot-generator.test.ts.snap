// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotGenerator should generate basic DOT output 1`] = `
"digraph "Test Graph" {
  graph [rankdir=LR, fontname="Arial", fontsize=12, overlap=scale, splines=true];
  node [shape=box, style="rounded,filled", fontname="Arial", fontsize=10];
  edge [color="#999999", fontname="Arial", fontsize=8];

  "app" [label="App", tooltip="App: App", shape="doubleoctagon", fillcolor="#ff9896"];
  "pageA" [label="PageA", tooltip="Page: PageA", shape="ellipse", fillcolor="#ffbb78"];
  "compB" [label="CompB", tooltip="Component: CompB", shape="box", fillcolor="#bcbd22"];
  "utilC" [label="UtilC", tooltip="Module: UtilC", shape="note", fillcolor="#aec7e8"];

  "app" -> "pageA" [tooltip="Structure", color="#333333", penwidth="1.5", style="solid"];
  "pageA" -> "compB" [tooltip="Structure", color="#333333", penwidth="1.5", style="solid"];
  "pageA" -> "utilC" [tooltip="Import", color="#1f77b4", style="dashed"];
  "compB" -> "utilC" [tooltip="Import", color="#1f77b4", style="dashed"];
}
"
`;
